signer/src/block_notifier.rs

```diff
@@ -0,0 +1,177 @@
+//! This module provides functionality for notifying about new blocks using the Electrum client.
+
+use crate::config::{BlockNotifierConfig, SETTINGS};
+use electrum_client::bitcoin::BlockHash;
+use electrum_client::{
+    Client, ConfigBuilder, ElectrumApi, Error as ElectrumError, HeaderNotification,
+};
+use futures::stream::{Stream, StreamExt};
+use std::sync::Arc;
+use tokio::sync::broadcast::{self, Receiver, Sender};
+use tokio::sync::Mutex;
+use tokio::task;
+use tokio::time::{interval, sleep, Duration};
+use tokio_stream::wrappers::BroadcastStream;
+use tracing::{debug, error};
+
+/// Block Notifier Errors
+#[derive(Debug, thiserror::Error, Clone)]
+pub enum Error {
+    /// Electrum error
+    #[error("electrum error: {0}")]
+    Electrum(Arc<electrum_client::Error>),
+    /// Lagged
+    #[error("BroadcastStream")]
+    BroadcastStream,
+}
+
+/// The `BlockNotifier` trait defines a method for subscribing to a stream of block headers.
+pub trait BlockNotifier {
+    /// Errors occurring during subscription or running the notifier.
+    type Error: std::error::Error;
+
+    /// Returns a stream of block headers.
+    fn subscribe(&self) -> impl Stream<Item = Result<BlockHash, Self::Error>> + Send;
+}
+
+/// A struct for polling Block Headers from Electrum client.
+pub struct ElectrumBlockNotifier {
+    client: Arc<Mutex<Client>>,
+    retry_interval: Duration,
+    max_retry_attempts: u32,
+    ping_interval: Duration,
+    subscribe_interval: Duration,
+}
+
+impl ElectrumBlockNotifier {
+    /// Creates a new instance of `ElectrumBlockNotifier` from config.
+    ///
+    /// # Returns
+    ///
+    /// A new instance of `ElectrumBlockNotifier`.
+    pub fn from_config(config: BlockNotifierConfig) -> Result<Self, ElectrumError> {
+        let server = &config.server;
+        let client_config = ConfigBuilder::new().build();
+        let client = Arc::new(Mutex::new(Client::from_config(server, client_config)?));
+        let retry_interval = Duration::from_secs(config.retry_interval);
+        let max_retry_attempts = config.max_retry_attempts;
+        let ping_interval = Duration::from_secs(config.ping_interval);
+        let subscribe_interval = Duration::from_secs(config.subscribe_interval);
+        Ok(ElectrumBlockNotifier {
+            client,
+            retry_interval,
+            max_retry_attempts,
+            ping_interval,
+            subscribe_interval,
+        })
+    }
+
+    /// Creates a new instance of `ElectrumBlockNotifier` from system config.
+    ///
+    /// # Returns
+    ///
+    /// A new instance of `ElectrumBlockNotifier`.
+    pub fn new() -> Result<Self, ElectrumError> {
+        let config = &SETTINGS.block_notifier;
+        Self::from_config(config.clone())
+    }
```

Comment:
If we take the config by reference:
```suggestion
    pub fn new() -> Result<Self, ElectrumError> {
        Self::from_config(&SETTINGS.block_notifier)
    }
```